### DISCUSSION TITLE
Allow only numeric input in Entry control?

### DISCUSSION THREAD

**Discussion by MattBDev** [dotnet/maui#9173] [Q&A]
I was wondering if it was possible to prevent non-numeric characters from being input in an Entry control? If that isn't possible, I want to delete text that the user entered that is invalid when the control looses focus, the completed event, or the return command is fired. I have been looking for an answer but I haven't been able to find one. As far as I can tell, it isn't possible to cancel the TextChanged event. Can anyone point me on the right track or provide an answer?

**Comment by EmberLynn** [comment]
Add this property to your Entry: Keyboard="Numeric"

**Reply by MattBDev** [reply]
Unfortunately that doesn't work. It only changes what an on-screen keyboard layout would display.

**Reply by EmberLynn** [reply]
I'm currently using an Android emulator for my app and I found that setting the property Keyboard="Numeric" only allows input of digits. Perhaps this is different on other platforms? Aside from that, you can use the Behaviors in the MAUI Community Toolkit to validate numbers. It won't prevent inputting non-valid characters into your field, however, you can use these attached behaviors to visually present an error to the user. 

https://docs.microsoft.com/en-us/dotnet/communitytoolkit/maui/behaviors/numeric-validation-behavior

**Reply by MattBDev** [reply]
You're right. After you mentioned this I loaded up my app on my Pixel 3 and it does limit input to just numbers on Android. I even tried inputting letters using a wireless keyboard and it didn't let me. It seems like .NET MAUI doesn't map Windows controls properly. I would expect if an entry has the property `Keyboard="Numeric"` it would use a NumberBox but that doesn't seem to be the case.

**Reply by Xyncgas** [reply]
🤦‍

**Reply by hansmbakker** [reply]
Indeed, see https://github.com/dotnet/maui/issues/9201

**Comment by MattBDev** [comment]
If it isn't possible to limit what characters are entered in an Entry control on Windows using .NET MAUI since it is using TextBox from WinUI is it possible to change the Entry control to use NumberBox from WinUI if the platform is Windows?

**Reply by JohnHSE** [reply]
If you are using MVVM and binding to a view model, the property of type int should enforce the value. Or crash. One of the other ;)

**Comment by MattBDev** [comment]
I'm still looking for help if anyone has any ideas?

**Reply by JohnHDev** [reply]
I replied with a working answer 9 days ago (from my work account). Binding to an int in your view model should work, and should only take a few minutes for you to try it.

**Reply by maztan** [reply]
One still can enter invalid characters

**Comment by Fred-Hudson-CST** [comment]
i have been look for a good way to do this for a while ... the following works but can definitely be improved 

`
private void MyEntry_Unfocused(object sender, FocusEventArgs e)
    {

        if (myValidator.IsValid)
        {
            ((MainPageVM)BindingContext).MyDouble = double.Parse(myEntry.Text);
            return;

        }
        else
        {
            myEntry.Text = ValidateAndConvert(myEntry.Text, ((MainPageVM)BindingContext).MyDouble.ToString(), myValidator.MinimumValue, myValidator.MaximumValue, myValidator.MinimumDecimalPlaces, myValidator.MaximumDecimalPlaces);
            myValidator.IsValid = true;
            ((MainPageVM)BindingContext).MyDouble = double.Parse(myEntry.Text);
            return;
        }

    }
`

`
protected string ValidateAndConvert(string? value, string oldValue, double minimumValue, double maximumValue, int minimumDecimalPlaces, int maximumDecimalPlaces)
    {
        ArgumentNullException.ThrowIfNull(value);

        // If the value cannot be parsed as a double, return the old value
        if (!double.TryParse(value, out var numeric))
        {
            return oldValue;
        }

        var decimalDelimiterIndex = value.IndexOf(CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator);
        var hasDecimalDelimiter = decimalDelimiterIndex >= 0;

        var decimalPlaces = hasDecimalDelimiter
            ? value.Substring(decimalDelimiterIndex + 1, value.Length - decimalDelimiterIndex - 1).Length
            : 0;

        // If there are too many decimal places, round the number
        if (decimalPlaces > maximumDecimalPlaces)
        {
            numeric = Math.Round(numeric, maximumDecimalPlaces);
        }

        // If there are too few decimal places, add the necessary amount and return it as a string
        if (decimalPlaces < minimumDecimalPlaces)
        {
            return numeric.ToString("F" + minimumDecimalPlaces, CultureInfo.CurrentCulture);
        }

        // After rounding, if the value is less than the minimum value, return the minimum value as a string
        if (numeric < minimumValue)
        {
            return minimumValue.ToString();
        }

        // After rounding, if the value is more than the maximum value, return the maximum value as a string
        if (numeric > maximumValue)
        {
            return maximumValue.ToString();
        }

        // If the value is within all specified ranges, return it as a string
        return numeric.ToString("F" + maximumDecimalPlaces, CultureInfo.CurrentCulture);
    }
`

**Comment by mikkelrj** [comment]
@MattBDev I know you asked this long ago, but here's my approach to solving this problem.  I am listening to TextChanged events on the Entry control and if the new value contains non-digits, I clean them from the value and set the modified value on the Entry.

Here's an example with an entry control for entering minutes (must be integer):

In page constructor:
```
        minutesEntry.TextChanged += MinutesEntry_TextChanged;
```

And the event handler: 

```
    private void MinutesEntry_TextChanged(object? sender, TextChangedEventArgs e)
    {
        if (!int.TryParse(e.NewTextValue, out _))
        {
            // clean out invalid characters
            minutesEntry.Text = new string(e.NewTextValue.Where(char.IsDigit).ToArray());
        }
    }
```

If you bind the entry to an integer property, like I have, you will get unhandled exceptions because the entry will try to assign the value before your handler is triggered. You can bind to an intermediate string-typed property to avoid this.

**Reply by zeplinn** [reply]
The problem with this solution is that you will fire the TextChanged event twice when you clean out invalid characters, meaning commands and bindable text properties will be called twice.
Next problem arise from the fact that to use this solution you need to limit your characters to fit the range of an int and the very odd behavior that "001" is considered valid input
also there is no reason to turn your string query into an array and you really should provide a null check
```
private void MinutesEntry_TextChanged(object? sender, TextChangedEventArgs e)
    {
       // if you do not have the null check you the code will get very brittle
        if (string.IsNullOrEmpty(e.Newvalue) && !int.TryParse(e.NewTextValue, out _))
        {
            // clean out invalid characters (no array casting)
            minutesEntry.Text = string.Concat(e.NewTextValue.Where(char.IsDigit));
        }
    }
```

**Comment by stephenquan** [comment]
Keyboard="Numeric" works well on Android. However, on Windows and iOS (particularly iPads), the user can enter invalid input. The `TextChanged` solution leads to a post-validation fix-up where the user will momentarily see the invalid text before it is corrected. We can trick or divert the user's attention to it, but, it still happens, e.g.

```c#
// MaskEntry.cs
public class MaskEntry : Entry
{
    public string Mask { get; set; } = @"^\-?\d*$";
    public MaskEntry()
    {
        TextChanged += async (s, e) =>
        {
            if (!string.IsNullOrEmpty(Mask) && !Regex.Match(e.NewTextValue, Mask).Success)
            {
                Opacity = 0.1;
                await Task.Delay(1);
                Opacity = 1;
                Text = e.OldTextValue;
            }
        };
    }
}
```
